neuro_py/ensemble/pairwise_bias_correlation.py

```diff
@@ -0,0 +1,395 @@
+from joblib import Parallel, delayed
+
+import nelpy as nel
+import numpy as np
+import sklearn
+import sklearn.metrics
+
+from neuro_py.io import loading
+from neuro_py.process import intervals
+from neuro_py.session.locate_epochs import (
+    compress_repeated_epochs,
+)
+from neuro_py.spikes import spike_tools
+
+
+def bias_matrix(
+    spike_times: np.ndarray,
+    neuron_ids: np.ndarray,
+    total_neurons: int,
+    fillneutral: float = 0.5
+) -> np.ndarray:
+    r"""
+    Compute the bias matrix for a given sequence of spikes.
+
+    Parameters
+    ----------
+    spike_times : numpy.ndarray
+        Spike times for the sequence, assumed to be sorted.
+    neuron_ids : numpy.ndarray
+        Neuron identifiers corresponding to `spike_times`.
+        Values should be integers between 0 and `total_neurons - 1`.
+    total_neurons : int
+        Total number of neurons being considered.
+    fillneutral : float, optional
+        Value to fill for neutral bias, by default 0.5
+
+    Returns
+    -------
+    numpy.ndarray
+        A bias matrix of size `(total_neurons, total_neurons)` where
+        each entry represents the bias between neuron pairs.
+
+    Notes
+    -----
+    The bias \( B_{ij} \) for neurons \( i \) and \( j \) is computed as:
+    \[
+    B_{ij} = \frac{nspikes_{ij}}{nspikes_i \cdot nspikes_j}
+    \]
+    where \( nspikes_{ij} \) is the count of spikes from neuron \( i \) occurring 
+    before spikes from neuron \( j \). If there are no spikes for either neuron,
+    the bias is set to 0.5 (neutral bias).
+    """
+    bias = np.empty((total_neurons, total_neurons))
+    np.fill_diagonal(bias, fillneutral)
+
+    # Create boolean masks for all neurons in advance
+    masks = [neuron_ids == i for i in range(total_neurons)]
+
+    for i in range(total_neurons):
+        spikes_i = spike_times[masks[i]]  # timestamps for neuron i
+        nspikes_i = spikes_i.size
+        for j in range(total_neurons):
+            if i != j:
+                spikes_j = spike_times[masks[j]]
+                nspikes_j = spikes_j.size
+
+                if nspikes_i > 0 and nspikes_j > 0:
+                    # Count how many times neuron i spikes before each neuron j spike
+                    nspikes_ij = np.searchsorted(
+                        spikes_i, spikes_j, side='right').sum()
+                    bias[i, j] = nspikes_ij / (nspikes_i * nspikes_j)
+                else:
+                    bias[i, j] = fillneutral  # Neutral bias if no spikes
+
+    return bias
+
+
+def bias_matrix_fast(
```

Comment:
do we need both normal and fast version? What is the trade off for fast? If nothing, we can just keep the fast

Reply:
I like the normal version for readability, easy to understand the code. If there's only fast version in the codebase, it can be quite hard to interpret.